src/lib/pubkey/curve448/curve448_utils/curve448_gf.h

```diff
@@ -0,0 +1,141 @@
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+#ifndef BOTAN_CURVE_448_GF_H_
+#define BOTAN_CURVE_448_GF_H_
+
+#include <botan/mem_ops.h>
+#include <botan/types.h>
+
+#include <array>
+#include <span>
+
+namespace Botan {
+
+/**
+ * This class represents a GF element in the field GF(2^448 - 2^224 - 1). Computations are
+ * performed using optimized operations as defined in the paper:
+ * "Reduction Modulo 2^448 - 2^224 - 1" by Kaushik Nath and Palash Sarkar
+ * (https://eprint.iacr.org/2019/1304).
+ *
+ * The representation of the field element is a 448-bit uint, stored in little-endian order
+ * as 7*64bit words. Note that the internal representation is not necessarily canonical, i.e.
+ * the value might be larger than the prime modulus. When calling the to_bytes() method, the
+ * canonical representation is returned.
+ */
+class Gf448Elem {
+   public:
+      /**
+       * @brief Construct a GF element from a 448-bit integer gives as 56 bytes @p x in
+       * little-endian order.
+       */
+      Gf448Elem(std::span<const uint8_t, 56> x);
+
+      /**
+       * @brief Construct a GF element from a 448-bit integer gives as 7 uint64_t words @p x in
+       * little-endian order.
+       */
+      Gf448Elem(std::span<const uint64_t, 7> data) { copy_mem(m_x, data); }
+
+      /**
+       * @brief Construct a GF element by passing the least significant 64 bits as a word.
+       * All other become zero.
+       */
+      Gf448Elem(uint64_t least_sig_word);
+
+      /**
+       * @brief Store the canonical representation of the GF element as 56 bytes in little-endian
+       * order.
+       *
+       * @param out The 56 byte output buffer.
+       */
+      void to_bytes(std::span<uint8_t, 56> out) const;
+
+      /**
+       * @brief Return the canonical representation of the GF element as 56 bytes in little-endian
+       * order.
+       */
+      std::array<uint8_t, 56> to_bytes() const;
+
+      /**
+       * @brief Swap this and other if b == true. Constant time for any b.
+       */
+      void ct_cond_swap(bool b, Gf448Elem& other);
+
+      /**
+       * @brief Set this to @p other if b is true. Constant time for any b.
+       */
+      void ct_cond_assign(bool b, const Gf448Elem& other);
+
+      Gf448Elem operator+(const Gf448Elem& other) const;
+
+      Gf448Elem operator-(const Gf448Elem& other) const;
+
+      Gf448Elem operator-() const;
+
+      Gf448Elem operator*(const Gf448Elem& other) const;
+
+      /**
+       * @brief out = this^2. Faster than operator*.
+       */
+      Gf448Elem square() const;
+
+      Gf448Elem operator/(const Gf448Elem& other) const;
+
+      /**
+       * @brief Compute the root of this element in the field.
+       *
+       * The root of a in GF(p) is computed as r = a^((p+1)/4) mod p.
+       * Note that the root is not unique, i.e. r and p-r are both roots.
+       *
+       * @return GfPElem The root of this element.
+       */
+      Gf448Elem root() const;
```

Comment:
I must admit I prefer the functions as members, since its an operation like any other and therefore belongs to the location all other ops are implemented. Also, its easier finding member functions using the IDE. I admit root(a+b) looks prettier than (a+b).root(), but both are quite understandable.

Reply:
Why do you use the `+` operator then? Under that logic, shouldn't you have an `auto add() -> Gf448Elem` method? 🤡 
Jokes aside: You have a point that free-standing functions are easier to find, as the IDE proposes them when you type `obj.`. But I'd argue that this internal class will be used in a write-once-read-many fashion, so we should optimize for readability and not so much for IDE affordance.

Albeit a bit chatty, we could have it both ways by adding free-standing wrappers like so:

```C++
inline Gf448Elem root(const Gf448Elem& x) {
   return x.root();
}
```

Additionally, the usage of free-standing functions is quite a common pattern in modern STL patterns. Such as `std::ranges::size()`, `std::advance(some_itr)`, and so on.

Reply:
> But I'd argue that this internal class will be used in a write-once-read-many fashion, so we should optimize for readability and not so much for IDE affordance.

Fair point. I'll change the interface.
However, I think the reason for freestanding functions like `std::ranges::size(some_range)` is to work on concepts.

Reply:
> I think the reason for freestanding functions like std::ranges::size(some_range) is to work on concepts.

Also extensibility. Imagine you have some third-party container type that doesn't provide (say) `begin()` and `end()` methods, but you want to use it in some abstract algorithm. When using free-standing functions, you -- as a user of the type -- can just provide free-standing overloads for the third-party type.

In any case: I just wanted to make the point that free-standing functions are nothing extraordinary these days. The technical reasons in the STL are indeed different, as you said.